python/ray/dag/compiled_dag_node.py

```diff
@@ -1075,6 +1124,221 @@ def _get_or_compile(
+
+        def _add_edge(in_node: GraphNode, out_node: GraphNode):
+            in_node.out_edges.add((out_node.idx, out_node.operation.type))
+            out_node.in_edges.add((in_node.idx, in_node.operation.type))
+
+        graph: Dict[int, Dict[DAGNodeOperationType, GraphNode]] = {}
+
+        from ray.dag import (
+            ClassMethodNode,
+            MultiOutputNode,
+        )
+
+        for _, executable_tasks in self.actor_to_executable_tasks.items():
+            prev_compute_node = None
+            for local_idx, exec_task in enumerate(executable_tasks):
+                idx = exec_task.idx
+                read_node = GraphNode(
+                    DAGNodeOperation(local_idx, DAGNodeOperationType.READ),
+                    idx,
+                    self.idx_to_task[idx].dag_node,
+                )
+                compute_node = GraphNode(
+                    DAGNodeOperation(local_idx, DAGNodeOperationType.COMPUTE),
+                    idx,
+                    self.idx_to_task[idx].dag_node,
+                )
+                write_node = GraphNode(
+                    DAGNodeOperation(local_idx, DAGNodeOperationType.WRITE),
+                    idx,
+                    self.idx_to_task[idx].dag_node,
+                )
+                _add_edge(read_node, compute_node)
+                _add_edge(compute_node, write_node)
+                if prev_compute_node is not None:
+                    _add_edge(prev_compute_node, compute_node)
+                prev_compute_node = compute_node
+                graph[idx] = {
+                    DAGNodeOperationType.READ: read_node,
+                    DAGNodeOperationType.COMPUTE: compute_node,
+                    DAGNodeOperationType.WRITE: write_node,
+                }
+
+        for idx, task in self.idx_to_task.items():
+            if not isinstance(task.dag_node, ClassMethodNode):
+                continue
+            for downstream_idx in task.downstream_node_idxs:
+                downstream_dag_node = self.idx_to_task[downstream_idx].dag_node
+                if isinstance(downstream_dag_node, MultiOutputNode):
+                    continue
+                _add_edge(
+                    graph[idx][DAGNodeOperationType.WRITE],
+                    graph[downstream_idx][DAGNodeOperationType.READ],
+                )
+
+        import heapq
+
+        actor_to_candidates = defaultdict(list)
+        for idx, node_dict in graph.items():
+            for _, node in node_dict.items():
+                if node.in_degree == 0:
+                    heapq.heappush(actor_to_candidates[node.actor_handle], node)
+
+        visited_nodes = set()
+
+        def _select_next_nodes():
+            """
+            Select the next nodes for topological sort. This function may return
+            multiple nodes if they are NCCL nodes. In that case, this function only
+            removes the NCCL write node, which is also the peek of a priority queue.
```

Comment:
Didn't quite understand this - if we are returning multiple nodes but only removing one, how do we know from the return values which one is being removed?

Reply:
We don't need to know which node in `next_nodes` is removed in this iteration. We just need to ensure that all nodes are added to `visited_nodes`, which should guarantee that a `DAGNodeOperation` is added to an execution schedule only once.

Reply:
Other nodes in `next_nodes` will be removed when they become the first node in `next_nodes`. Hence, `_select_next_nodes` will be called `num_graph_nodes` times.

Reply:
Move the discussion to https://github.com/ray-project/ray/pull/46911#discussion_r1708199574.